src/children.js

```diff
@@ -29,3 +31,27 @@ export const properties = children.properties = {
 
 // Old JSEP versions
 properties.LogicalExpression = properties.BinaryExpression;
+
+/**
+ * Replace a child node with a new node by updating its parent
+ * @param {object} child the node to replace
+ * @param {object} newChild the replacement node
+ */
+export function replace (child, newChild) {
+	const parent = parents.get(child);
+	if (!parent) {
+		return;
+	}
+	const childProperties = properties[parent.type];
+	for (const prop of childProperties) {
+		if (parent[prop] === child) {
+			parent[prop] = newChild;
+		}
+		else if (Array.isArray(parent[prop])) {
+			const index = parent[prop].indexOf(child);
+			if (index !== -1) {
+				parent[prop][index] = newChild;
+			}
+		}
+	}
+}
```

Comment:
Pretty sure we have abstractions we can use here rather than reimplement the logic of looking through children in yet another place. If our traversal abstractions are not flexible enough to support this, they should *become* flexible enough.

Reply:
I don't think we do currently, so I'll work on that too

Reply:
SGTM. Let's discuss potential API designs before implementing!

Reply:
A potential solution could be to create an alt version of the `children` function, except instead of returning an array of children, it returns an array of objects in the form
```js
{
    property: string;
    index?: number;
    node: object;
}
```
Then this function could be easily called to iterate over the children while maintaining the metadata needed to set children.

Reply:
As general advice for these things, don't think about it from the perspective of "how to expose this data?", think of it like "I, as a user, need this functionality, what would feel natural for me to use?"

Some of my thoughts so far:
- I think thinking about it in terms of `index` is [overfit](https://bootcamp.uxdesign.cc/overfitting-and-the-problem-with-use-cases-337d9f4bf4d7). Index is not really a separate concept, it's just that sometimes you need one step to get to the child (`parent[prop]`) and sometimes more (`parent[prop][secondProp]`). You could use the concept of a *path* from the parent to the child, which usually has one property and sometimes two (and potentially more in the future, though it's unlikely). That could even generalize to finding the path between any two nodes (if any), which seems like exactly the kind of low level primitive that is more broadly useful. While a total generalization would be challenging to implement in an equally performant way, if it's explicit which node is the ancestor and which the descendant and it only allows going up/down and not both, I think it would be comparable. Then once we have the path, the rest is trivial.
- `walk()` seems somewhat related, as it also gets children and passes `(node, property, parent)` to the callback. It doesn't do quite what we want though because it flattens the array and only gives us one property, and also because it recurses on the whole subtree, whereas here we just need one step. *But what if it (or a version of it) could not have these constraints?*
- `children()` as you point out also seems related, Perhaps there could be a version that returns a `Map` of nodes to paths.

Reply:
I like your first idea of having a general path function, something like `path(ancestor, descendant)` which returns an array of keys (strings for props or numbers for array indices) used to get there. What do you think of that?

Reply:
SGTM!

Reply:
I will make a separate PR/issue for it